sktime/annotation/hmm.py

```diff
@@ -0,0 +1,314 @@
+        # state at that stage
+        trans_prob = np.zeros((num_states, num_obs))
+        trans_prob[:, 0] = np.log(initial_probs)
+
+        # trans_id is the index of the state that would have been the most
+        # likely preceeding state.
+        trans_id = np.zeros((num_states, num_obs), dtype=np.int32)
+
+        # use Vertibi Algorithm to fill in trans_prob and trans_id:
+        for i in range(1, num_obs):
+            # use log probabilities to try to keep nums reasonable -Inf
+            # means 0 probability
+            paths = np.zeros((num_states, num_states))
+            for j in range(num_states):
+                paths[j, :] += trans_prob[:, i - 1]  # adds prev trans_prob column-wise
+                paths[:, j] += np.log(emi_probs[:, i])  # adds log(probs_sub) row-wise
+            paths += np.log(
+                transition_prob_mat
+            )  # adds log(transition_prob_mat) element-wise
+            trans_id[:, i] = np.argmax(paths, axis=0)
+            trans_prob[:, i] = np.max(paths, axis=0)
+
+        if np.any(np.isinf(trans_prob[:, -1])):
+            raise ValueError("Change parameters, the distribution doesn't work")
+
+        return trans_prob, trans_id
+
+    @classmethod
+    def _make_emission_probs(
+        cls, emission_funcs: list, observations: np.ndarray
+    ) -> np.ndarray:
+        """Calculate the prob each obs comes from each hidden state."""
+        # assign emission probabilities from each state to each position:
+
+        emi_probs = np.zeros(shape=(len(emission_funcs), len(observations)))
+        for state_id, emission_func in enumerate(emission_funcs):
+            if isinstance(emission_func, tuple):
+                emission_func = emission_func[0]
+                kwargs = emission_func[1]
+                emi_probs[state_id, :] = np.array(
+                    [emission_func(x, **kwargs) for x in observations]
+                )
+            else:
+                emi_probs[state_id, :] = np.array(
+                    [emission_func(x) for x in observations]
+                )
+        return emi_probs
+
+    def _hmm_viterbi_label(self) -> np.array:
+        """Assign hidden state ids to all observations based on most likely path.
+
+        Parameters
+        ----------
+        self: (HMM) an HMM instance which already has already had trans_id and trans_mat
+            calculated.
+
+        Returns
+        -------
+        hmm_fit: (np.ndarray) an mx1 array where m is the length of the X (obs).
+            each entry in the array is an int representing a hidden id state
+            that has been assigned to that observation.
+        """
+        hmm_fit = np.zeros(self.num_obs)
+        # Now we trace backwards and find the most likely path:
+        max_inds = np.zeros(self.num_obs, dtype=np.int32)
+        max_inds[-1] = np.argmax(self.trans_prob[:, -1])
+        hmm_fit[-1] = self.states[max_inds[-1]]
+        for index in reversed(list(range(1, self.num_obs))):
```

Comment:
This might read less intuitive but we could use just `range` with all its params `(start, stop, step)` here, like 
`for index in range(self.num_obs, 0, -1)`, 
what do you think?

Reply:
I think that reads a lot better actually!  Thanks for the suggestion!